### DISCUSSION TITLE
Definition extention

### DISCUSSION THREAD

**Discussion by dudk** [cue-lang/cue#2074] [Q&A] [answered]
Hello and huge thanks for such a wonderful project!

I'm just making my first steps with CUE and  I'm trying to achieve the following:

1. Have a base definition with a set of fields;
2. Allow extending base definition with additional fields;
3. Additional fields might refer to fields from the base definition.

Overall, I got it working with using the following pattern:

```
#Base2: {
    attrs: {...}
}

#Def2: #Base2 & {
    attrs: {
        a: number
        b: number
    }
}

#definitionExtension2: #Def2 & {
    attrs: {
        c: attrs.a * attrs.b
    }
}

val2: #definitionExtension2 & {
    attrs: { 
        a: 2
        b: 3
    }
}
```

Outputs:

```
val2: {
	attrs: {
		a: 2
		c: 6
		b: 3
	}
}
```

What I want to understand if it's possible to get rid of `attrs` inner struct and have it working with something like this:

```
#Base1: {...}

#Def1: #Base1 & {
    a: number
    b: number
}

#definitionExtension1: #Def1 & {
    //c: a * b 
}
```

Which manifests the following error: 
```
#definitionExtension1.c: reference "a" not found:
    -:9:8
```

It's confusing because I expect both cases to work since in the end they both are extending definitions, but seems like I don't understand difference in semantics of these examples.

[Link](https://tip.cuelang.org/play/?id=CjEKkYIlxRt#cue@export@cue) to the playground with the code above.

**Comment by slewiskelly** [comment, marked as answer]
You need to use an [alias](https://cuelang.org/docs/references/spec/#aliases) to reference these fields:

```cue
#Base1: {...}

#Def1: #Base1 & {
	a: number
	b: number
}

#definitionExtension1: X=#Def1 & {
	c: X.a * X.b
}

d: #definitionExtension1 & {
	a: 2
	b: 3
}
```

Further to that, I would recommend against using an open struct, instead opting for [embedding](https://cuelang.org/docs/references/spec/#embedding):

```cue
#Base: {
	a: number
}

#Def: {
	#Base

	b: number
}

#Ext: X={
	#Def

	c: X.a + X.b
}

e: #Ext & {
	a: 1
	b: 2
}
```

**Reply by dudk** [reply]
Thank you for a swift reply @slewiskelly! It makes sense to me.
